

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Command Line Utilites &mdash; RNAchrom 0 documentation</title>
  

  
  

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  
    <link rel="top" title="RNAchrom 0 documentation" href="index.html"/>
        <link rel="next" title="RNAchrom’s functions" href="packageFunctions.html"/>
        <link rel="prev" title="Welcome to RNAchrom’s documentation!" href="index.html"/> 

  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/modernizr/2.6.2/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        <a href="index.html" class="fa fa-home"> RNAchrom</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
        
            <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="">Command Line Utilites</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#annotate-bam-files">Annotate BAM files</a></li>
<li class="toctree-l2"><a class="reference internal" href="#compute-correlation-between-number-of-links-and-number-of-histone-peaks">Compute correlation between number of links and number of histone peaks</a></li>
<li class="toctree-l2"><a class="reference internal" href="#count-long-range-interactions">Count long range interactions</a></li>
<li class="toctree-l2"><a class="reference internal" href="#remove-overlapps">Remove overlapps</a></li>
<li class="toctree-l2"><a class="reference internal" href="#filter-out-reads-showing-known-splice-junctions">Filter out reads showing known splice-junctions</a></li>
<li class="toctree-l2"><a class="reference internal" href="#compute-specificity-of-blind-links">Compute specificity of blind-links</a></li>
<li class="toctree-l2"><a class="reference internal" href="#detect-strong-interactions">Detect strong interactions</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="packageFunctions.html">RNAchrom&#8217;s functions</a></li>
</ul>

        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">RNAchrom</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>Command Line Utilites</li>
      <li class="wy-breadcrumbs-aside">
        
          <a href="_sources/commandLineTools.txt" rel="nofollow"> View page source</a>
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            
  <div class="section" id="command-line-utilites">
<span id="commandlinetools"></span><h1>Command Line Utilites<a class="headerlink" href="#command-line-utilites" title="Permalink to this headline">¶</a></h1>
<div class="section" id="annotate-bam-files">
<h2>Annotate BAM files<a class="headerlink" href="#annotate-bam-files" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text"><div class="highlight"><pre>usage: annotateBAM [-h] [-b BFILE] [-g GFILE] [-r RFILE] [-o OFILE]

Annotate both mates of paired-end alignments

optional arguments:
  -h, --help  show this help message and exit
  -b BFILE    BAM file(s). A single aligment file contains all paired-end
              mates, or a comma separated list of 2 files, one per each set of
              mates (useful when mate origin is important).
  -g GFILE    GTF file. Annotation file with biotype information
  -r RFILE    BED file. Annotation file with repeats information
  -o OFILE    STR. Name of output file. Default = &quot;output&quot;
</pre></div>
</div>
</div>
<div class="section" id="compute-correlation-between-number-of-links-and-number-of-histone-peaks">
<h2>Compute correlation between number of links and number of histone peaks<a class="headerlink" href="#compute-correlation-between-number-of-links-and-number-of-histone-peaks" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text"><div class="highlight"><pre>usage: corrWind [-h] [-l LFILE] [-e EXON] [-w WINDOW] [-t LINKTYPE] [-d DIST]
                [-c CFILE] [-f FOLDER] [-o OFILE]

Computes correlation between number of aimer-target links and number of ChIP-
seq/DNase peaks

optional arguments:
  -h, --help   show this help message and exit
  -l LFILE     BED file. Paired-end alignment reads are presente on each line.
               Output of annotateBAM.py
  -e EXON      STR BOOL. &quot;True&quot;, for checking that a mate should overlap and
               exon to be counted as outgoing link of a window (if both mates
               are on the same window, at least one of them should overlap an
               exon). &quot;False&quot;, if want to count all links.
  -w WINDOW    INT. Window size (nt) when splitting genome.
  -t LINKTYPE  STR. Link type: aware, blind.
  -d DIST      INT. Distance between mates to be consider not selfligating.
  -c CFILE     TAB file. Chromosome size information
  -f FOLDER    STR. Folder where the ChIP-seq/DNase peak files are stored
  -o OFILE     STR. Name of output file.
</pre></div>
</div>
</div>
<div class="section" id="count-long-range-interactions">
<h2>Count long range interactions<a class="headerlink" href="#count-long-range-interactions" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text"><div class="highlight"><pre>usage: countHits [-h] [-a AFILE] [-g GFILE] [-d DISTANCE] [-o OFILE]

Counts the number of RNA and DNA mates per gene (including introns). Only
mates where the RNA-ends overlap exons are use.

optional arguments:
  -h, --help   show this help message and exit
  -a AFILE     BED file. Output of annotateBAM.py, where the left and right
               ends correspond to DNA and RNA, respectively
  -g GFILE     GTF file. Annotation file with biotype information
  -d DISTANCE  INT. Minimum distance (in nucleotides) between RNA-DNA ends to
               be considered as a long range interaccion. Default = 2000 b
  -o OFILE     STR. Name of output file.
</pre></div>
</div>
</div>
<div class="section" id="remove-overlapps">
<h2>Remove overlapps<a class="headerlink" href="#remove-overlapps" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text"><div class="highlight"><pre>usage: removeOverlapps [-h] [-i IFILE] [-o OFILE]

Remove overlapping reads.

optional arguments:
  -h, --help  show this help message and exit
  -i IFILE    BED file. Coordenate-sorted output of annotateBAM.py.
  -o OFILE    BED file. Name of output file. Same format as input file.
</pre></div>
</div>
</div>
<div class="section" id="filter-out-reads-showing-known-splice-junctions">
<h2>Filter out reads showing known splice-junctions<a class="headerlink" href="#filter-out-reads-showing-known-splice-junctions" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text"><div class="highlight"><pre>usage: sj [-h] [-b BFILE] [-d DIST] [-o OFILE]

Get interesting read&#39;s IDs (printed to STDOUT). That&#39;s, reads that: (1) have
spliced junctions concordant with the annotated genome, and (2) have a mate
farther apart than a given distance

optional arguments:
  -h, --help  show this help message and exit
  -b BFILE    BAM file(s). A single aligment file contains all paired-end
              mates.
  -d DIST     INT. Minimum distance between paired-end mates
  -o OFILE    STR. Name of output file.
</pre></div>
</div>
</div>
<div class="section" id="compute-specificity-of-blind-links">
<h2>Compute specificity of blind-links<a class="headerlink" href="#compute-specificity-of-blind-links" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text"><div class="highlight"><pre>usage: specificity [-h] [-a AFILE] [-b BFILE] [-e EXON] [-w WINDOW] [-d DIST]
                   [-c CFILE] [-o OFILE]

Computes fold change of blind-links on regions targeted by aware-lings vs
blind-links targeting everywhere

optional arguments:
  -h, --help  show this help message and exit
  -a AFILE    BED file. Aware-links file. The first and second mates
              correspond to DNA and RNA annotated using annotateBAM.py
  -b BFILE    BED file. Blind-links file. annotated using annotateBAM.py
  -e EXON     STR BOOL. &quot;True&quot;, for checking that a mate should overlap and
              exon to be counted as outgoing link of a window (if both mates
              are on the same window, at least one of them should overlap an
              exon). &quot;False&quot;, if want to count all links.
  -w WINDOW   INT. Window size (nt) when splitting genome.
  -d DIST     INT. Distance between mates to be consider not selfligating.
  -c CFILE    TAB file. Chromosome size information
  -o OFILE    STR. Name of output file.
</pre></div>
</div>
</div>
<div class="section" id="detect-strong-interactions">
<h2>Detect strong interactions<a class="headerlink" href="#detect-strong-interactions" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text"><div class="highlight"><pre>usage: detectLinks [-h] [-a AFILE] [-b BFILE] [-g GFILE] [-w WINDSIZE]
                   [-d DIST] [-o OFILE]

Find strong aimer-target relations. It uses aware-links (know and inferred) as
seed to determine target genomic regions(TGR). A TGR is build by overlapping
the windows of target DNA-mates. Conversely, exons are used as aimer genomic
regions

optional arguments:
  -h, --help   show this help message and exit
  -a AFILE     TAB file. TAB separated file with aware mates&#39; information
               (output of annotateBAM.py).
  -b BFILE     TAB file. TAB separated file with blind mates&#39; information
               (output of annotateBAM.py).
  -g GFILE     GTF file.
  -w WINDSIZE  INT. Window size (nt) to build around aware links to determine
               target regions.
  -d DIST      INT. Distance between mates used as lower threshold to call
               them non self-ligating.
  -o OFILE     STR. Name of output file
</pre></div>
</div>
</div>
</div>


          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="packageFunctions.html" class="btn btn-neutral float-right" title="RNAchrom’s functions"/>Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="index.html" class="btn btn-neutral" title="Welcome to RNAchrom’s documentation!"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014, mrivas.
    </p>
  </div>

  <a href="https://github.com/snide/sphinx_rtd_theme">Sphinx theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>
</footer>
        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>