

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Example case &mdash; RNAchrom 0 documentation</title>
  

  
  

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  
    <link rel="top" title="RNAchrom 0 documentation" href="index.html"/>
        <link rel="next" title="Command Line Utilites" href="commandLineTools.html"/>
        <link rel="prev" title="Prerequisites and installation" href="installation.html"/> 

  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/modernizr/2.6.2/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        <a href="index.html" class="fa fa-home"> RNAchrom</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
        
            <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Prerequisites and installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="installation.html#prerequisites">Prerequisites</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html#installation-on-linux">Installation on Linux</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Example case</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#data-summary">Data summary</a></li>
<li class="toctree-l2"><a class="reference internal" href="#reads-alignment">Reads Alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="#interesting-cases-rna-over-dna-enriched-genes">Interesting cases: RNA over DNA enriched genes</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="commandLineTools.html">Command Line Utilites</a><ul>
<li class="toctree-l2"><a class="reference internal" href="commandLineTools.html#annotate-bam-files">Annotate BAM files</a></li>
<li class="toctree-l2"><a class="reference internal" href="commandLineTools.html#compute-correlation-between-number-of-links-and-number-of-histone-peaks">Compute correlation between number of links and number of histone peaks</a></li>
<li class="toctree-l2"><a class="reference internal" href="commandLineTools.html#count-long-range-interactions">Count long range interactions</a></li>
<li class="toctree-l2"><a class="reference internal" href="commandLineTools.html#remove-overlapps">Remove overlapps</a></li>
<li class="toctree-l2"><a class="reference internal" href="commandLineTools.html#filter-out-reads-showing-known-splice-junctions">Filter out reads showing known splice-junctions</a></li>
<li class="toctree-l2"><a class="reference internal" href="commandLineTools.html#compute-specificity-of-blind-links">Compute specificity of blind-links</a></li>
<li class="toctree-l2"><a class="reference internal" href="commandLineTools.html#detect-strong-interactions">Detect strong interactions</a></li>
</ul>
</li>
</ul>

        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">RNAchrom</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>Example case</li>
      <li class="wy-breadcrumbs-aside">
        
          <a href="_sources/exampleCase.txt" rel="nofollow"> View page source</a>
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            
  <div class="section" id="example-case">
<span id="examplecase"></span><h1>Example case<a class="headerlink" href="#example-case" title="Permalink to this headline">¶</a></h1>
<div class="section" id="data-summary">
<h2>Data summary<a class="headerlink" href="#data-summary" title="Permalink to this headline">¶</a></h2>
<p>TwoStep1024 and TwoStep1219</p>
</div>
<div class="section" id="reads-alignment">
<h2>Reads Alignment<a class="headerlink" href="#reads-alignment" title="Permalink to this headline">¶</a></h2>
<p>We aligned the paired-end reads with STAR [Dobin2013], allowing up to 2 mismatches and restricting the intron size (for mates spanning a spliced junction) to less than 10k nt. Reads were divided into two sets: one where at at least one mate of each pair contains the linker sequence, and another where both mates of each pair doesn’t contain the linker sequence (see Figure <a href=".html#alignment_diagram"></a><tt class="xref num docutils literal"><span class="pre">#alignment_diagram</span></tt>). The first set was further split into RNA and DNA mates (based on the direction of the linker sequence; reads were the linker sequence was on the flanking regions were discarded), and paired-end aligned to produce aware-links. The second set was paired-end aligned to produce blind-links. In all cases duplicates were removed. For libraries TwoStep1024 and TwoStep1219, this resulted in 18M links for both libraries, from where 7k and 5k were aware links (see Table1)</p>
<div class="figure" id="alignment-diagram">
<a class="reference internal image-reference" href="https://132.239.135.28/public/RNAchrom/files/exampleCase/alignment_diagram.svg"><img src="https://132.239.135.28/public/RNAchrom/files/exampleCase/alignment_diagram.svg" width="90%" /></a>
<p class="caption">Figure 1: Construction of blind and aware-links. Paired-end fastq files (fastq1,2) were aligned using STAR (Dobin 2013) allowing for up to 2 mismatches. On one hand, reads not containing the linker sequence resulted in aligned (mates’ distance &lt;600k) and chimeric (mates’s distance &gt;600k, or on different chromosomes) alignments, both of them used to build blind-links. On the other hand, reads containing the linker sequence in were splitted into DNA and RNA fragments (determined by the removed linker sequence orientation) and then aligned to build aware-links.</p>
</div>
<table border="1" class="docutils">
<caption>Table1: Number of links per library. In parenthesis the % of links overlapping exons in at least one mate for all links and on the RNA mate of aware-links.</caption>
<colgroup>
<col width="17%" />
<col width="17%" />
<col width="17%" />
<col width="17%" />
<col width="17%" />
<col width="17%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Sample</th>
<th class="head"># all-links</th>
<th class="head"># aware-links</th>
<th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>TwoSteps1024</td>
<td>18</td>
<td>053</td>
<td>131 (9.53%)</td>
<td>6</td>
<td>898 (6.07%)</td>
</tr>
<tr class="row-odd"><td>TwoSteps1219</td>
<td>18</td>
<td>048</td>
<td>508 (8.30%)</td>
<td>5</td>
<td>015 (4.62%)</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="interesting-cases-rna-over-dna-enriched-genes">
<h2>Interesting cases: RNA over DNA enriched genes<a class="headerlink" href="#interesting-cases-rna-over-dna-enriched-genes" title="Permalink to this headline">¶</a></h2>
<p>We use aware-links to determine candidate cases for aimers, i.e., genes that interact with distal genomic regions. For the genomic region of each gene, interesting cases were judged based on their RNA over DNA-ends ratio. Compared to DNA-mates, RNA-mates should be more concentrated on aimer genes as these contain both:</p>
<ol class="arabic simple">
<li>links reaching other genes (targets of the current gene) and,</li>
<li>links of self-ligation (DNA of the current gene binding its owns nascent mRNA).</li>
</ol>
<p>Exceptions to this would be DNA regions that are targets of other aimers. To test whether this is the case we computed the distributions of RNA and DNA ends only using aware-links overlapping coding regions. The histograms of RNA and DNA-ends per gene (Figure <a href=".html#matesoverlength"></a><tt class="xref num docutils literal"><span class="pre">#matesOverLength</span></tt>) are not only clearly different (negative skewness Figure <a href=".html#ratios"></a><tt class="xref num docutils literal"><span class="pre">#ratios</span></tt>), but the mean value of RNA-ends per gene is bigger than its equivalent on DNA, meaning that most genes contain more RNA than DNA ends. Thus, there are clear cases where the number of RNA is bigger than the number of DNA-mates. We focus on those for candidates. To stress the RNA over DNA enrichment we computed their ratio per genomic region (Figure <a href=".html#ratios"></a><tt class="xref num docutils literal"><span class="pre">#ratios</span></tt>) and select the top 25%.</p>
<div class="figure" id="matesoverlength">
<a class="reference internal image-reference" href="https://132.239.135.28/public/RNAchrom/files/exampleCase/matesOverLength.svg"><img src="https://132.239.135.28/public/RNAchrom/files/exampleCase/matesOverLength.svg" width="90%" /></a>
<p class="caption">Figure 2: Distribution of DNA and RNA mates per gene. All values were normalized by gene length.</p>
</div>
<div class="figure" id="rations">
<a class="reference internal image-reference" href="https://132.239.135.28/public/RNAchrom/files/exampleCase/ratios.svg"><img src="https://132.239.135.28/public/RNAchrom/files/exampleCase/ratios.svg" width="90%" /></a>
<p class="caption">Figure 3: Distribution of RNA over DNA reads per gene. All values were normalzied by gene length. Red line correspond to 75% quantile.</p>
</div>
<p>This resulted in 4 candidates protein coding genes that are shared among replicates (see Table2).</p>
<table border="1" class="docutils">
<caption>Table 2: Candidate genes shared among replicates.</caption>
<colgroup>
<col width="14%" />
<col width="14%" />
<col width="14%" />
<col width="14%" />
<col width="14%" />
<col width="14%" />
<col width="14%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Gene</th>
<th class="head">Length (nt)</th>
<th class="head">Biotype</th>
<th class="head"># DNA mates</th>
<th class="head"># RNA mates</th>
<th class="head">Enrichment</th>
<th class="head">&nbsp;</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>ENSMUSG00000058537</td>
<td>3</td>
<td>408</td>
<td>protein coding</td>
<td>0</td>
<td>1</td>
<td>0.693</td>
</tr>
<tr class="row-odd"><td>ENSMUSG00000075014</td>
<td>755</td>
<td>protein coding</td>
<td>0</td>
<td>4</td>
<td>1.61</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>ENSMUSG00000078942</td>
<td>36</td>
<td>554</td>
<td>protein coding</td>
<td>0</td>
<td>1</td>
<td>0.693</td>
</tr>
<tr class="row-odd"><td>ENSMUSG00000018906</td>
<td>31</td>
<td>571</td>
<td>protein coding</td>
<td>0</td>
<td>1</td>
<td>0.693</td>
</tr>
<tr class="row-even"><td>ENSMUSG00000064358</td>
<td>784</td>
<td>protein coding</td>
<td>0</td>
<td>2</td>
<td>1.1</td>
<td>&nbsp;</td>
</tr>
</tbody>
</table>
</div>
</div>


          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="commandLineTools.html" class="btn btn-neutral float-right" title="Command Line Utilites"/>Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="installation.html" class="btn btn-neutral" title="Prerequisites and installation"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014, mrivas.
    </p>
  </div>

  <a href="https://github.com/snide/sphinx_rtd_theme">Sphinx theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>
</footer>
        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>