

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Example case &mdash; RNAchrom 0 documentation</title>
  

  
  

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  
    <link rel="top" title="RNAchrom 0 documentation" href="index.html"/>
        <link rel="next" title="Command Line Utilites" href="commandLineTools.html"/>
        <link rel="prev" title="Prerequisites and installation" href="installation.html"/> 

  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/modernizr/2.6.2/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        <a href="index.html" class="fa fa-home"> RNAchrom</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
        
            <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Prerequisites and installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="installation.html#prerequisites">Prerequisites</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html#installation-on-linux">Installation on Linux</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Example case</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#data-summary">Data summary</a></li>
<li class="toctree-l2"><a class="reference internal" href="#interesting-cases-rna-end-over-dna-end-enriched-genes">Interesting cases: RNA-end over DNA-end enriched genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="#inferred-aware-links">Inferred aware-links</a></li>
<li class="toctree-l2"><a class="reference internal" href="#bibliography">Bibliography</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="commandLineTools.html">Command Line Utilites</a><ul>
<li class="toctree-l2"><a class="reference internal" href="commandLineTools.html#annotate-bam-files">Annotate BAM files</a></li>
<li class="toctree-l2"><a class="reference internal" href="commandLineTools.html#compute-correlation-between-number-of-links-and-number-of-histone-peaks">Compute correlation between number of links and number of histone peaks</a></li>
<li class="toctree-l2"><a class="reference internal" href="commandLineTools.html#count-long-range-interactions">Count long range interactions</a></li>
<li class="toctree-l2"><a class="reference internal" href="commandLineTools.html#remove-overlapps">Remove overlapps</a></li>
<li class="toctree-l2"><a class="reference internal" href="commandLineTools.html#filter-out-reads-showing-known-splice-junctions">Filter out reads showing known splice-junctions</a></li>
<li class="toctree-l2"><a class="reference internal" href="commandLineTools.html#compute-specificity-of-blind-links">Compute specificity of blind-links</a></li>
<li class="toctree-l2"><a class="reference internal" href="commandLineTools.html#detect-strong-interactions">Detect strong interactions</a></li>
</ul>
</li>
</ul>

        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">RNAchrom</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>Example case</li>
      <li class="wy-breadcrumbs-aside">
        
          <a href="_sources/exampleCase.txt" rel="nofollow"> View page source</a>
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            
  <div class="section" id="example-case">
<span id="examplecase"></span><h1>Example case<a class="headerlink" href="#example-case" title="Permalink to this headline">¶</a></h1>
<p>Here, we&#8217;ll use RNAchrom tools to analyze RNA-chromatin data. In their raw form, RNA-chromatin libraries are paired-end sequencing libraries. Each read pair correspond to an RNA-chromatin link, where one mate comes from the RNA aimer and the other from its chromatin target. For brevity, we&#8217;ll call these as RNA-mate and DNA-mate, respectively. In the same fashion, we&#8217;ll call the gene sources of the RNA-mates as aimer genes.</p>
<div class="section" id="data-summary">
<h2>Data summary<a class="headerlink" href="#data-summary" title="Permalink to this headline">¶</a></h2>
<p>We generated two RNA-chromatin libraries on mouse E14: TwoStep1024, and TwoStep1219 (see Table 1). For each library, read pairs were divided into two sets:</p>
<ul class="simple">
<li><strong>Aware pairs</strong>, where at least one mate contained the linker sequence. Based on the direction of the linker sequence, each mate could be resolved as RNA-end or DNA-end.</li>
<li><strong>Blind pairs</strong>, where linker sequences were absent in both mates. Therefore, there was not direct way to map mates into RNA-end and DNA-ends.</li>
</ul>
<p>As the RNA-mates were expected to span splice junctions, we used STAR &#8211;as splice tolerant aligner <a class="reference internal" href="#dobin2013" id="id1">[1]</a>&#8211; to map the reads against mm9. For all sets, we allowed up to 2 mismatches, restricted the intron size to less than 10k nt (Figure <a href="#alignment-diagram"> 1</a>). For both libraries most blind pairs could be aligned (&gt;75%). Aware pairs, on the other hand, were less mappable (alignment rates ~49%) mainly because after removing the linker sequence we ended up with shorter mates sequences.</p>
<div class="figure" id="alignment-diagram">
<a class="reference internal image-reference" href="https://132.239.135.28/public/RNAchrom/files/exampleCase/alignment_diagram.svg"><img src="https://132.239.135.28/public/RNAchrom/files/exampleCase/alignment_diagram.svg" width="90%" /></a>
<p class="caption">Figure 1: Construction of blind and aware-links. Paired-end fastq files (fastq1,2) were aligned using STAR (Dobin 2013) allowing for up to 2 mismatches. On one hand, reads not containing the linker sequence resulted in aligned (mates’ distance &lt;10k) and chimeric (mates’s distance &gt;600k, or on different chromosomes) alignments, both of them used to build blind-links. On the other hand, reads containing the linker sequence in were splitted into DNA and RNA fragments (determined by the removed linker sequence orientation) and then aligned to build aware-links.</p>
</div>
<p>After the alignment, pairs whose both mates were uniquely mapped were considered <strong>RNA-chromatin links</strong>. Aware or bind pairs with one or two mates having multi-hits were discarded.  To avoid PCR-amplification biases, we removed duplicates. This resulted in ~16M links in both libraries, among them were 6k and 4k aware links, respectively (see Table1).</p>
<table border="1" class="docutils">
<caption>Table1: Number of links per library. To compute the % of links overlapping exons, we requiered at least one mate on a pair should overlap an exon.</caption>
<colgroup>
<col width="20%" />
<col width="20%" />
<col width="20%" />
<col width="20%" />
<col width="20%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Sample</th>
<th class="head"># aware pairs (% uniquely aligned)</th>
<th class="head"># blind pairs (% uniquely aligned)</th>
<th class="head"># aware links (% overlapping exons)</th>
<th class="head">blind links (% overlapping exons)</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>TwoSteps1024</td>
<td>105,657 (48.88)</td>
<td>29,505,340 (75.01)</td>
<td>5,996 (9.9)</td>
<td>15,752,812 (9.52)</td>
</tr>
<tr class="row-odd"><td>TwoSteps1219</td>
<td>71,981 (48.53)</td>
<td>31,765,882 (78.84)</td>
<td>4,357 (7.64)</td>
<td>16,522,999 (7.98)</td>
</tr>
</tbody>
</table>
<p><strong>Inter-chromosomal links</strong> were found only among-blind links: 509 and 368 for libraries TwoSteps1024 an TwoSteps1219, respectively.</p>
</div>
<div class="section" id="interesting-cases-rna-end-over-dna-end-enriched-genes">
<h2>Interesting cases: RNA-end over DNA-end enriched genes<a class="headerlink" href="#interesting-cases-rna-end-over-dna-end-enriched-genes" title="Permalink to this headline">¶</a></h2>
<p>We used aware-links to determine the genes interacting with distal genomic regions. For the genomic region of each gene, interesting cases were judged based on their RNA-ends over DNA-ends ratio. The basic idea is that compared to DNA-ends, RNA-ends should be more concentrated on aimer genes as these contain both:</p>
<ol class="arabic simple">
<li>links reaching other genes (targets of the current gene) and,</li>
<li>links of self-ligation (DNA of the current gene binding its owns nascent mRNA).</li>
</ol>
<p>Exceptions to this would be genes that are themselves targets of other aimers. To test whether this is a generalized case we computed the distributions of RNA and DNA-ends only using aware-links overlapping coding regions. The histograms of RNA and DNA-ends per gene (Figure <a href="#mates-over-length"> 2</a>) are not only clearly different (negative skewness Figure <a href="#ratios"> 3</a>), but the mean value of RNA-ends per gene is bigger than its equivalent on DNA, meaning that most genes contain more RNA than DNA ends. Thus, there are clear cases where the number of RNA-ends is bigger than the number of DNA-ends. We focused on those as candidates. To stress the RNA over DNA enrichment we computed their ratio per genomic region (Figure <a href="#ratios"> 3</a>) and select the top 25%.</p>
<div class="figure" id="mates-over-length">
<a class="reference internal image-reference" href="https://132.239.135.28/public/RNAchrom/files/exampleCase/matesOverLength.svg"><img src="https://132.239.135.28/public/RNAchrom/files/exampleCase/matesOverLength.svg" width="90%" /></a>
<p class="caption">Figure 2: Distribution of DNA and RNA mates per gene. All values were normalized by gene length.</p>
</div>
<div class="figure" id="ratios">
<a class="reference internal image-reference" href="https://132.239.135.28/public/RNAchrom/files/exampleCase/ratios.svg"><img src="https://132.239.135.28/public/RNAchrom/files/exampleCase/ratios.svg" width="90%" /></a>
<p class="caption">Figure 3: Distribution of RNA over DNA reads per gene. All values were normalzied by gene length. Red line correspond to 75% quantile.</p>
</div>
<p>This resulted in 2 candidates genes that are shared among replicates (see Table 2).</p>
<table border="1" class="docutils">
<caption>Table 2: Candidate genes shared among replicates. DNA and RNA-mates correspond to the number of supporting mates on the genomic region of the respective gene (including introns).</caption>
<colgroup>
<col width="14%" />
<col width="14%" />
<col width="14%" />
<col width="14%" />
<col width="14%" />
<col width="14%" />
<col width="14%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Gene ensemble ID</th>
<th class="head">gene name</th>
<th class="head">Length (nt)</th>
<th class="head">Biotype</th>
<th class="head"># DNA mates</th>
<th class="head"># RNA mates</th>
<th class="head">Enrichment</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>ENSMUSG00000078942</td>
<td>Naip6</td>
<td>36,554</td>
<td>protein coding</td>
<td>0</td>
<td>1</td>
<td>0.693</td>
</tr>
<tr class="row-odd"><td>ENSMUSG00000018906</td>
<td>P4ha2</td>
<td>31,571</td>
<td>protein coding</td>
<td>0</td>
<td>1</td>
<td>0.693</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="inferred-aware-links">
<h2>Inferred aware-links<a class="headerlink" href="#inferred-aware-links" title="Permalink to this headline">¶</a></h2>
<p>Since the aware-links are a small fraction of the blind-links, and the previous enriched cases are statistically insignificant (low number of supporting links), we resorted to an inference method of aware-links. Among blind-links, we looked for mates spanning known splicing junctions. When one end of a pair meets this criteria it was inferred that it corresponds to the RNA-end. By asscociation, the other mate was considered as the DNA-mate. As example, see GM108000&#8217;s links at the <a class="reference external" href="http://epigenomegateway.wustl.edu/browser/?genome=mm9&amp;session=KNZWb8e2Mq&amp;statusId=178461723">WashU genome browser</a>. There, it can be seen that this gene contains only 1 aware-link (from library TwoSteps1219), still, many blind-links spanning known splicing junctions (track *allSJ) support the connection of this gene with its upstream neighbor Gm10801-201. What&#8217;s more, this inferred links seems to be supported by the remaining blind-links.</p>
</div>
<div class="section" id="bibliography">
<h2>Bibliography<a class="headerlink" href="#bibliography" title="Permalink to this headline">¶</a></h2>
<p id="bibtex-bibliography-exampleCase-0"><table class="docutils citation" frame="void" id="dobin2013" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id1">[1]</a></td><td>Alexander Dobin, Carrie&nbsp;A Davis, Felix Schlesinger, Jorg Drenkow, Chris Zaleski, Sonali Jha, Philippe Batut, Mark Chaisson, and Thomas&nbsp;R Gingeras. STAR: ultrafast universal RNA-seq aligner.. <em>Bioinformatics (Oxford, England)</em>, 29(1):15–21, January 2013. <a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/early/2012/10/25/bioinformatics.bts635">URL: http://bioinformatics.oxfordjournals.org/content/early/2012/10/25/bioinformatics.bts635</a>, <a class="reference external" href="http://dx.doi.org/10.1093/bioinformatics/bts635">doi:10.1093/bioinformatics/bts635</a>.</td></tr>
</tbody>
</table>
</p>
</div>
</div>


          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="commandLineTools.html" class="btn btn-neutral float-right" title="Command Line Utilites"/>Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="installation.html" class="btn btn-neutral" title="Prerequisites and installation"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014, mrivas.
    </p>
  </div>

  <a href="https://github.com/snide/sphinx_rtd_theme">Sphinx theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>
</footer>
        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>