<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>RNAchrom’s functions &mdash; RNAchrom 0 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="RNAchrom 0 documentation" href="index.html" />
    <link rel="prev" title="Command Line Utilites" href="commandLineTools.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="commandLineTools.html" title="Command Line Utilites"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">RNAchrom 0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="module-_RNAchrom">
<span id="rnachrom-s-functions"></span><h1>RNAchrom&#8217;s functions<a class="headerlink" href="#module-_RNAchrom" title="Permalink to this headline">¶</a></h1>
<dl class="function">
<dt id="_RNAchrom.annotate">
<tt class="descclassname">_RNAchrom.</tt><tt class="descname">annotate</tt><big>(</big><em>almnt</em>, <em>genes</em>, <em>exon</em>, <em>repeat</em><big>)</big><a class="headerlink" href="#_RNAchrom.annotate" title="Permalink to this definition">¶</a></dt>
<dd><p>Annotate an alignment</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>almnt</strong> &#8211; Alignment, HTSeq.GenomicInterval</li>
<li><strong>genes</strong> &#8211; Genes, gene[genomic_iv]=set([geneID1,geneID2,...])</li>
<li><strong>exon</strong> &#8211; Exons, exon[genomic_iv]=set([&#8220;exon&#8221;])</li>
<li><strong>repeat</strong> &#8211; Repeats regions, HTSeq.GenomicArrayOfSets[genomic_iv]=set([repeatID])</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">Annotation string, formed as: &#8220;[gene_annot1,gene_annot2,...|.]|[exon,intron|.]|[repeat_annot|.]&#8221;. Dot (.) denotes absence of annotation.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="_RNAchrom.bed2links">
<tt class="descclassname">_RNAchrom.</tt><tt class="descname">bed2links</tt><big>(</big><em>bFile</em>, <em>chromLength</em>, <em>windSize</em>, <em>checkExon</em>, <em>dist</em>, <em>linkType</em><big>)</big><a class="headerlink" href="#_RNAchrom.bed2links" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert DNA-RNA links from BED to GenomicArrayOfSets format</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>bFile</strong> &#8211; TXT file. Annotated file of mates (output of annotateBAM.py).</li>
<li><strong>chromLength</strong> &#8211; DICT. Dictionary of chromosome lengths.</li>
<li><strong>windSize</strong> &#8211; INT. Window size in which the genome is divided.</li>
<li><strong>checkExon</strong> &#8211; BOOL. Whether to check if RNA-mates overlapp exons.</li>
<li><strong>dist</strong> &#8211; INT. Distance between used as lower threshold to determine self-ligation.</li>
<li><strong>linkType</strong> &#8211; STR. Type of mates: &#8220;aware&#8221;, or &#8220;blind&#8221;.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">DICT[DICT]=INT. Dictionary of target per aimer.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="_RNAchrom.bed2peaks">
<tt class="descclassname">_RNAchrom.</tt><tt class="descname">bed2peaks</tt><big>(</big><em>File</em>, <em>chromLength</em>, <em>windSize</em><big>)</big><a class="headerlink" href="#_RNAchrom.bed2peaks" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert DNA-RNA links from BED to GenomicArrayOfSets format</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>File</strong> &#8211; BED file. File name where ChIP-seq peaks are stored.</li>
<li><strong>chromLength</strong> &#8211; DICT. Dictionary of chromosome lengths.</li>
<li><strong>windSize</strong> &#8211; INT. Window size in which the genome is divided.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">DICT. Number of peaks per genomic window.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="_RNAchrom.buildLinks">
<tt class="descclassname">_RNAchrom.</tt><tt class="descname">buildLinks</tt><big>(</big><em>awareMates</em>, <em>blindMates</em>, <em>windSize</em>, <em>genesID_IV</em>, <em>oFile</em><big>)</big><a class="headerlink" href="#_RNAchrom.buildLinks" title="Permalink to this definition">¶</a></dt>
<dd><p>Count hits of aware and blind-mates. It determine target areas based on aware-mates.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>awareMates</strong> &#8211; DICT. Dictionary of aware mates.</li>
<li><strong>blindMates</strong> &#8211; DICT. Dictionary of blind mates.</li>
<li><strong>windSize</strong> &#8211; INT. Window size of expanded target mates.</li>
<li><strong>genes</strong> &#8211; DICT. Dictionary of genes&#8217; genomic intervals, where keys are gene ids and values HTSeq.GenomicIntervals.</li>
<li><strong>oFile</strong> &#8211; STR. Name of output TAB separated file, where results will be stored. The output format is: geneName geneid_iv target_iv awareCounts awareSD blindCounts blindSD.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="_RNAchrom.chromLength">
<tt class="descclassname">_RNAchrom.</tt><tt class="descname">chromLength</tt><big>(</big><em>chromFile</em><big>)</big><a class="headerlink" href="#_RNAchrom.chromLength" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a DICT of chromosome lengths.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>chromFile</strong> &#8211; TXT file. TAB separated file where first and second columns correspond to chromosome name, and chromosome size (nt), respectively.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">DICT. Dictionary where keys and values are chrom names and chrom sizes, respectively.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="_RNAchrom.counts">
<tt class="descclassname">_RNAchrom.</tt><tt class="descname">counts</tt><big>(</big><em>aFile</em>, <em>distance</em>, <em>geneIV_ID</em><big>)</big><a class="headerlink" href="#_RNAchrom.counts" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the counts of long range interactions for each gene</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>aFile</strong> &#8211; Annotated file of aware links (output of annotateBAM.py).</li>
<li><strong>genIV_ID</strong> &#8211; DICT. Dictionary where keys and values are gene genomic intervals and gene ids, respectively.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Oaran distance:</th><td class="field-body"><p class="first">INT. Distance between mates used a lower threshold to be call as not self-ligating.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">[DICT,DICT]. Array of dictionaries, corresponding to the count of DNA and RNA per gene id (key), respectively.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="_RNAchrom.expandIv">
<tt class="descclassname">_RNAchrom.</tt><tt class="descname">expandIv</tt><big>(</big><em>iv</em>, <em>windSize</em><big>)</big><a class="headerlink" href="#_RNAchrom.expandIv" title="Permalink to this definition">¶</a></dt>
<dd><p>Expand genomic interval</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>iv</strong> &#8211; HTSeq.GenomicInterval. Genomic interval to be expanded</li>
<li><strong>windSize</strong> &#8211; INT. Final size of the expanded iv</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">HTSeq.GenomicInterval. Genomic interval centered at iv and with length windSize</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="_RNAchrom.formatOutput">
<tt class="descclassname">_RNAchrom.</tt><tt class="descname">formatOutput</tt><big>(</big><em>first</em>, <em>second</em>, <em>annotFirst</em>, <em>annotSecond</em><big>)</big><a class="headerlink" href="#_RNAchrom.formatOutput" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes alignments&#8217; information and creates an string fit to be printed.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>first</strong> &#8211; Genomic interval, HTSeq.GenomicInterval</li>
<li><strong>second</strong> &#8211; Genomic interval, HTSeq.GenomicInterval</li>
<li><strong>annotFirst</strong> &#8211; Annotation of <em>first</em></li>
<li><strong>annotSecond</strong> &#8211; Annotation of <em>second</em></li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">String of annotations, &#8220;first.chrom, first.start, first.end, first.strand, first.annot, first.jMField, second.chrom,...&#8221;</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="_RNAchrom.geneExonIV">
<tt class="descclassname">_RNAchrom.</tt><tt class="descname">geneExonIV</tt><big>(</big><em>gtfFile</em><big>)</big><a class="headerlink" href="#_RNAchrom.geneExonIV" title="Permalink to this definition">¶</a></dt>
<dd><p>Stores genomic intervals of genes (including introns) and exons in dictionaries.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>gtfFile</strong> &#8211; GTF file.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">[gene,exon] GenomicArrayOfSets, gene[genomic_iv]=set([geneID1,geneID2,...]), GenomicArrayOfSets, exon[genomic_iv]=&#8221;exon&#8221;</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="_RNAchrom.genes">
<tt class="descclassname">_RNAchrom.</tt><tt class="descname">genes</tt><big>(</big><em>gtfFile</em><big>)</big><a class="headerlink" href="#_RNAchrom.genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Get overlap genomic regions per each annotated feature in the annotation file</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>gtfFile</strong> &#8211; GTF file.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">[DICT,DICT]. Dictionaries of gene id (key) gene iv (value), and gene iv (key) and gene id (value), respectively.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="_RNAchrom.getBioType">
<tt class="descclassname">_RNAchrom.</tt><tt class="descname">getBioType</tt><big>(</big><em>gtfFile</em><big>)</big><a class="headerlink" href="#_RNAchrom.getBioType" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a dictionary of biotypes per gene.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>gtfFile</strong> &#8211; GTF file.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">DICT. Keys and values are gene id and biotypes, respectively</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="_RNAchrom.ivReg">
<tt class="descclassname">_RNAchrom.</tt><tt class="descname">ivReg</tt><big>(</big><em>iv</em>, <em>chromLength</em>, <em>windSize</em><big>)</big><a class="headerlink" href="#_RNAchrom.ivReg" title="Permalink to this definition">¶</a></dt>
<dd><p>Regularization of intervals. The input iv is transformed to a canonical region. This to avoid partial overlapping, and to get well defined aimer and target windows.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>iv</strong> &#8211; HTSeq.GenomicInterval. Genomic interval to be regularized.</li>
<li><strong>chromLength</strong> &#8211; DICT. Dictionary of chromosome lengths.</li>
<li><strong>windSize</strong> &#8211; INT. Window size in which the genome is divided.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">HTSeq.GenomicInterval. Regularized genomic interval containing input iv.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="_RNAchrom.lineToIv">
<tt class="descclassname">_RNAchrom.</tt><tt class="descname">lineToIv</tt><big>(</big><em>line</em>, <em>dist</em><big>)</big><a class="headerlink" href="#_RNAchrom.lineToIv" title="Permalink to this definition">¶</a></dt>
<dd><p>Create iv for DNA and RNA mates</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>line</strong> &#8211; STR. Line from annotatedBAM.py output.</li>
<li><strong>dist</strong> &#8211; INT. Distance between used as lower threshold to determine self-ligation.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">ARRAY. [HTSeq.GenomicInterval,HTSeq.GenomicInterval,BOOL,BOOL,BOOL]. Fist mate genomic interval, second mate genomic interval,</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="_RNAchrom.loadMates">
<tt class="descclassname">_RNAchrom.</tt><tt class="descname">loadMates</tt><big>(</big><em>annotFile</em>, <em>Type</em>, <em>dist</em><big>)</big><a class="headerlink" href="#_RNAchrom.loadMates" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a dictionary of mates. If input mates are blind, it infers aware-mates based on splicing information.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>annotFile</strong> &#8211; TXT file. Annotation file (output of annotateBAM.py).</li>
<li><strong>Type</strong> &#8211; STR. Type of links on annotFile. Could be: &#8220;aware&#8221; (second mate correspond to RNA-end), or &#8220;blind&#8221; (it&#8217;s not known what mates correspond to the RNA-end).</li>
<li><strong>dist</strong> &#8211; INT. Distance (nt) between mates to be considered not self-ligating (mates closer than this value are discarded).</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">DICT. Dictionary of mates, where the key is the geneID of the aimer, and the value, the interval of the target region. If Type==&#8221;aware&#8221;, it output a single DICT of aware mates. If type==&#8221;blind&#8221;, it output and array of two DICTs, the first are the inferred aware-mates, and the second the blind-mates.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="_RNAchrom.repeats">
<tt class="descclassname">_RNAchrom.</tt><tt class="descname">repeats</tt><big>(</big><em>rFile</em><big>)</big><a class="headerlink" href="#_RNAchrom.repeats" title="Permalink to this definition">¶</a></dt>
<dd><p>Stores genomic intervals of repeats as HTSeq.GenomicArrayOfSets.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>rFile</strong> &#8211; BED file with the name of repeats on 4th column.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">HTSeq.GenomicArrayOfSets[genomic_iv]=set([repeatID])</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="_RNAchrom.std">
<tt class="descclassname">_RNAchrom.</tt><tt class="descname">std</tt><big>(</big><em>readName_ivs</em><big>)</big><a class="headerlink" href="#_RNAchrom.std" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes standard deviation of a list of genomic intervals</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>iv_array</strong> &#8211; ARRAY[ HTSeq.GenomicInterval ]. Array of genomic intervals.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">INT. Standard deviation of input genomic intervals.</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="commandLineTools.html"
                        title="previous chapter">Command Line Utilites</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/packageFunctions.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="commandLineTools.html" title="Command Line Utilites"
             >previous</a> |</li>
        <li><a href="index.html">RNAchrom 0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2014, mrivas.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2b2.
    </div>
  </body>
</html>